core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -1643,7 +1646,10 @@ class SparkContext(config: SparkConf) extends Logging {
   def killExecutors(executorIds: Seq[String]): Boolean = {
     schedulerBackend match {
       case b: ExecutorAllocationClient =>
-        b.killExecutors(executorIds, replace = false, force = true).nonEmpty
+        require(executorAllocationManager.isEmpty,
```

Comment:
This is a developer api, so probably ok, but this is a change in behavior. Is it just not possible to support this with dynamic allocation?

Reply:
What would calling this mean with dynamic allocation on?  Note this api explicitly says its meant to adjust resource usage downwards.  If you've got just one executor, and then you kill it, should your app sit with 0 executors?  Or even if you've got 10 executors, and you kill one -- when is dynamic allocation allowed to bump the total back up?  I can't think of useful clear semantics for this

(though this is not necessary to fix the bug, I could pull this out and move to a discussion in a new jira)

Reply:
I'm not sure why you'd use this with dynamic allocation, but it's been possible in the past. It's probably ok to change this though.

Reply:
Hi, @squito , I'm quite questioned about the cases:
>  If you've got just one executor, and then you kill it, should your app sit with 0 executors?

if app sit with 0 executors, then pending tasks increase, which lead to `ExecutorAllocationManager` increases target number of executors. So, app will not always sit with 0 executors.

> Or even if you've got 10 executors, and you kill one -- when is dynamic allocation allowed to bump the total back up?

for this case, to be honest, I really do not get your point. But, it must blame my poor English.

And, what will happens if we use this method without `ExecutorAllocationManager `? Or do we really need adjust TargetNumExecutors (set `adjustTargetNumExecutors  = true` below) if we are not using `ExecutorAllocationManager `?

see these several lines in `killExecutors()`:
```
if (adjustTargetNumExecutors) {
  requestedTotalExecutors = math.max(requestedTotalExecutors - executorsToKill.size, 0)
  ...
  doRequestTotalExecutors(requestedTotalExecutors)
}
```
Set `adjustTargetNumExecutors  = true` will change `requestedTotalExecutors` . And IIUC, `requestedTotalExecutors ` is only used in dynamic allocation mode. So, if we are not  using `ExecutorAllocationManager `, allocation client will request `requestedTotalExecutors = 0`  number of executors to cluster manager (this is really terrible). But, actually, app without `ExecutorAllocationManager ` do not have a limit requesting executors (in default).

Actually, I think this series methods, including `killAndReplaceExecutor `,  `requestExecutors`, etc, are designed with dynamic allocation mode. And if we still want use these methods while app do not use `ExecutorAllocationManager`, we should not change `requestedTotalExecutors `, or even not request cluster manager with a specific number.

WDYT?

Reply:
My point in general is that the semantics of combining `SparkContext.killExecutors()` (which is a publicly visible function, which the end user can call) with dynamic allocation aren't well defined, and I have no idea what the behavior really should be.  I was giving some examples of weird behavior.  

>> If you've got just one executor, and then you kill it, should your app sit with 0 executors?

> if app sit with 0 executors, then pending tasks increase, which lead to ExecutorAllocationManager increases target number of executors. So, app will not always sit with 0 executors.

Thats true -- but only *when* pending tasks increase.  But if you've got 0 executors, how do you expect pending tasks to increase?  That would only happen when another taskset gets submitted, but with no executors your spark program will probably just be blocked.

In the other case, I'm just trying to point out strange interactions between user control and dynamic allocation control.  Imagine this sequence:

Dynamic Allocation: 1000 tasks, so 1000 executors
User: I only want 10 executors, so let me tell spark to kill 990 of them
...
... another taskset is submitted to add 1 more task ...
Dynamic Allocation: 1001 tasks, so 1001 executors
User: ??? I set the target to 10 executors, what happened?


>  So, if we are not using ExecutorAllocationManager, allocation client will request requestedTotalExecutors = 0 number of executors to cluster manager (this is really terrible)

hmm, from a quick look, I think you're right.  it doesn't seem that using `sc.killExecutors()` doesn't make sense even with dynamic allocation off.  I think `CoarseGrainedSchedulerBackend` should actually initiliaze `requestedTotalExecutors` with `SchedulerBackendUtils.getInitialTargetExecutorNumber`

Reply:
Hi @squito , thanks for your reply.

> but only *when* pending tasks increase.

`ExecutorAllocationManager ` will check pending (or backlog) tasks periodically. So, we do not have to wait for *increment* actually.

And for `Dynamic Allocation` & `User` case, yeah, that's hard to define. 

And I checked `SchedulerBackendUtils.getInitialTargetExecutorNumbe`, it set `DEFAULT_NUMBER_EXECUTORS` = 2. But, this is not consistent with `Master`, which set `executorLimit` to `Int.MaxValue` if we are not under dynamic allocation mode. Maybe we can just init `requestedTotalExecutors ` with `Int.MaxValue`(only when we are not under dynamic allocation mode). 
Or, we do not call `doRequestTotalExecutors` if we call `requestExecutors` or `killExecutors`, except `requestTotalExecutors`(only when we are not under dynamic allocation mode).

Reply:
@squito any thoughts?